Framework/Algorithms/src/FFTPreProcessing.cpp

```diff
@@ -0,0 +1,228 @@
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/ListValidator.h"
+#include "MantidKernel/StartsWithValidator.h"
+
+#include <cmath>
+#include <numeric>
+#include <vector>
+
+namespace Mantid {
+namespace Algorithms {
+
+using namespace Kernel;
+using API::Progress;
+using std::size_t;
+
+// Register the class into the algorithm factory
+DECLARE_ALGORITHM(FFTPreProcessing)
+
+/** Initialisation method. Declares properties to be used in algorithm.
+ *
+ */
+void FFTPreProcessing::init() {
+  declareProperty(make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input),
+                  "The name of the input 2D workspace.");
+  declareProperty(make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+                      "OutputWorkspace", "", Direction::Output),
+                  "The name of the output 2D workspace.");
+  std::vector<int> empty;
+  declareProperty(
+      Kernel::make_unique<Kernel::ArrayProperty<int>>("Spectra", empty),
+      "The workspace indices to preProcess.");
+  declareProperty(
+      "ApodizationFunction", "None",
+      boost::make_shared<Mantid::Kernel::StringListValidator>(
+          std::vector<std::string>{"None", "Bartlett", "Lorentz", "Connes",
+                                   "Cosine", "Gaussian", "Welch"}),
+      "The apodization function to apply to the data");
+  declareProperty("DecayConstant", 1.5,
+                  "The decay constant for the apodization function.");
+  auto mustBePositive = boost::make_shared<Kernel::BoundedValidator<int>>();
+  mustBePositive->setLower(0);
+  declareProperty(
+      "Padding", 0, mustBePositive->clone(),
+      "The amount of padding to add to the data,"
+      "it is the number of multiples of the data set."
+      "i.e 0 means no padding and 1 will double the number of data points.");
+}
+
+/** Executes the algorithm
+ *
+ */
+void FFTPreProcessing::exec() {
+  std::vector<int> spectra = getProperty("Spectra");
+
+  // Get original workspace
+  API::MatrixWorkspace_const_sptr inputWS = getProperty("InputWorkspace");
+  auto numSpectra = inputWS->getNumberHistograms();
+  // Create output workspace with same dimensions as input
+  API::MatrixWorkspace_sptr outputWS = getProperty("OutputWorkspace");
+  if (inputWS != outputWS) {
+    outputWS = API::WorkspaceFactory::Instance().create(inputWS);
+  }
+
+  // Share the X values
+  for (size_t i = 0; i < static_cast<size_t>(numSpectra); ++i) {
+    outputWS->setSharedX(i, inputWS->sharedX(i));
+  }
+
+  // No spectra specified = process all spectra
+  if (spectra.empty()) {
+    spectra = std::vector<int>(numSpectra);
+    std::iota(spectra.begin(), spectra.end(), 0);
+  }
+
+  Progress prog(this, 0.0, 1.0, numSpectra + spectra.size());
+  if (inputWS != outputWS) {
+
+    // Copy all the Y and E data
+    PARALLEL_FOR_IF(Kernel::threadSafe(*inputWS, *outputWS))
+    for (int64_t i = 0; i < int64_t(numSpectra); ++i) {
```

Comment:
Does it make sense to keep unprocessed spectra? 
* If spectra are padded they get larger while the unprocessed ones have their old size. This is wrong in a `MatrixWorkspace`
* If there is no padding but the spectra are histograms. The processed ones become point data but the rest are histograms.